File: example-system/sim/sensor_node/packet.py
"""Uplink packet encode/decode (ICD section 4, SN-REQ-004..006, 009). Mirrors src/packet.c."""

import struct
from dataclasses import dataclass

SYNC = 0xA5
PACKET_LEN = 16
FLAG_FAULT = 0x01
FLAG_IMU_REINIT = 0x02

_HEADER = struct.Struct(">BBHhhhhh")  # sync, flags, seq, acc_x, acc_y, acc_z, gyro_z, temp_cc


def crc16_ccitt_false(data: bytes) -> int:
    """CRC-16/CCITT-FALSE: poly 0x1021, init 0xFFFF, no reflection, no xorout."""
    crc = 0xFFFF
    for byte in data:
        crc ^= byte << 8
        for _ in range(8):
            if crc & 0x8000:
                crc = ((crc << 1) ^ 0x1021) & 0xFFFF
            else:
                crc = (crc << 1) & 0xFFFF
    return crc


@dataclass(frozen=True)
class Packet:
    seq: int
    acc_x: int
    acc_y: int
    acc_z: int
    gyro_z: int
    temp_cc: int
    flags: int = 0

    @property
    def fault(self) -> bool:
        return bool(self.flags & FLAG_FAULT)


def build_packet(p: Packet) -> bytes:
    if not 0 <= p.seq <= 0xFFFF:
        raise ValueError("seq out of range")
    if p.flags & ~0x03:
        raise ValueError("reserved flag bits must be zero")
    body = _HEADER.pack(SYNC, p.flags, p.seq, p.acc_x, p.acc_y, p.acc_z, p.gyro_z, p.temp_cc)
    return body + struct.pack(">H", crc16_ccitt_false(body))


def parse_packet(raw: bytes) -> Packet:
    if len(raw) != PACKET_LEN:
        raise ValueError(f"packet must be {PACKET_LEN} bytes, got {len(raw)}")
    sync, flags, seq, ax, ay, az, gz, temp = _HEADER.unpack(raw[:14])
    if sync != SYNC:
        raise ValueError("bad sync byte")
    (crc,) = struct.unpack(">H", raw[14:])
    if crc != crc16_ccitt_false(raw[:14]):
        raise ValueError("crc mismatch")
    return Packet(seq=seq, acc_x=ax, acc_y=ay, acc_z=az, gyro_z=gz, temp_cc=temp, flags=flags)


def next_seq(seq: int) -> int:
    return (seq + 1) & 0xFFFF

File: example-system/src/packet.c
#include "packet.h"

uint16_t crc16_ccitt_false(const uint8_t *data, size_t len) {
    uint16_t crc = 0xFFFF;
    for (size_t i = 0; i < len; i++) {
        crc ^= (uint16_t)(data[i] << 8);
        for (int b = 0; b < 8; b++) {
            crc = (crc & 0x8000) ? (uint16_t)((crc << 1) ^ 0x1021) : (uint16_t)(crc << 1);
        }
    }
    return crc;
}

static void put_u16(uint8_t *dst, uint16_t v) {
    dst[0] = (uint8_t)(v >> 8);
    dst[1] = (uint8_t)(v & 0xFF);
}

static uint16_t get_u16(const uint8_t *src) { return (uint16_t)((src[0] << 8) | src[1]); }

bool packet_build(const packet_t *p, uint8_t out[PACKET_LEN]) {
    if (p->flags & (uint8_t)~(FLAG_FAULT | FLAG_IMU_REINIT)) return false;
    out[0] = PACKET_SYNC;
    out[1] = p->flags;
    put_u16(&out[2], p->seq);
    put_u16(&out[4], (uint16_t)p->acc_x);
    put_u16(&out[6], (uint16_t)p->acc_y);
    put_u16(&out[8], (uint16_t)p->acc_z);
    put_u16(&out[10], (uint16_t)p->gyro_z);
    put_u16(&out[12], (uint16_t)p->temp_cc);
    put_u16(&out[14], crc16_ccitt_false(out, 14));
    return true;
}

bool packet_parse(const uint8_t *raw, size_t len, packet_t *out) {
    if (len != PACKET_LEN || raw[0] != PACKET_SYNC) return false;
    if (get_u16(&raw[14]) != crc16_ccitt_false(raw, 14)) return false;
    out->flags = raw[1];
    out->seq = get_u16(&raw[2]);
    out->acc_x = (int16_t)get_u16(&raw[4]);
    out->acc_y = (int16_t)get_u16(&raw[6]);
    out->acc_z = (int16_t)get_u16(&raw[8]);
    out->gyro_z = (int16_t)get_u16(&raw[10]);
    out->temp_cc = (int16_t)get_u16(&raw[12]);
    return true;
}
